### PR TITLE
Introduce TaggedEvent

### PR DESCRIPTION
**Pull request by hernanponcedeleon**
This PR introduces a general `TaggedEvent` whose semantics are purely defined by the tags it has. 
Currently this is only used to implement fences, but there is another use case in the pipeline related to vulkan memory model.

The `Fence` class has almost no code now. I considered completely removing it, but since fences are such a basic and used concept in WMM, I decided to keep it.

I also refactor `FenceBase` to avoid repeated code between this class and `Fence`.

### CODE REVIEW

File: dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModel.java
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModel.java:365]: This code certainly violates the fact that `TaggedEvents` should be identified by their tags only (semantics-wise).
IIRC, this is done like this right now because we treat `fencerel(X)` as a base relation rather than a short-cut for `po;[X];po`.
I would put a `FIXME` there so we fix this eventually. You should double check what `data.isFence()` does and if it still applies for general `TaggedEvents`!

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/TaggedEvent.java
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/TaggedEvent.java:12]: This is needed for many WMM that use the name in the cat file, e.g. `MFENCE`. Do you want to have a parameter to the constructor to control if the name should be added or not?

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/TaggedEvent.java
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/TaggedEvent.java:28]: I don't think we should do that because that is mixing two things. Whatever is chosen for a name should be informative enough to understand the event semantics.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/visitors/EventVisitor.java
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/visitors/EventVisitor.java:38]: I don't get what you mean. `visitFence` is gone because I removed `Fence` (push pending). But it seems you also want to remove `visitTaggedEvent` ...

File: dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModel.java
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModel.java:365]: I don't fully understand what your concern is.
```
public boolean isFence() { return event.hasTag(FENCE); }
```
thus the event is being identified by its tag (and if that invariant is satisfied, I don't see the problem with the cast)

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/visitors/EventVisitor.java
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/visitors/EventVisitor.java]: We can keep `visitTaggedEvent` but I'm not sure if that is really a good idea, because the whole point is that you never need to know the class but only its tags (i.e., any check like `instanceof TaggedEvent` or a cast to `TaggedEvent` is HIGHLY suspicious).
But I think it is fine to keep `visitTaggedEvent` because it does have one parameter that is not just a tag, namely its name. That might be enough to justify the existence of dedicated visitor methods for it.

File: dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModel.java
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModel.java:365]: The problem is that the code assumes not only that the event has a fence tag but also that it is a `TaggedEvent`.
What if you had something like a `FencedLoad` that acts as a fence and a load simultaneously? In fact, acquire/release memory accesses fit into that picture (they just don't have the `F` tag).

Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModel.java:365]: ok, now I got it. I agree, but the old code had the same issue. For now I will simply add a FIXME

**Comment by ThomasHaas** [comment]
I need to think about what the best way is to handle `TaggedEvent`. Just from seeing the resulting code, I really dislike its usage.
That being said, I also think that our old usage of `Fence` was only fine as long as one could assume the invariant that fences are disjoint from all other visible events. I am not sure how well this invariant applies to `TaggedEvent`...

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/TaggedEvent.java
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/TaggedEvent.java]: I think the `EventFactory` methods can just put the name (e.g., `MFENCE`) twice, once as the display name and once as a tag. No need to have an extra flag.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java:110]: I would return `GenericVisibleEvent(name, name, Tag.FENCE)` to make it semantically equivalent with the previous version. Then you do not need to change any of the other fence factory methods.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java:465]: If you fix the `newFence` method, you can reduce this back to `return newFence("DMB.ISH")`.
Same for all other fence methods

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/common/FenceBase.java
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/event/common/FenceBase.java:11]: The comment is outdated.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/common/FenceBase.java
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/event/common/FenceBase.java:17]: Why do we require a memory order now? I think fences without a `mo` are reasonable.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/LogThreadStatistics.java
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/LogThreadStatistics.java:42]: Maybe we should change this to `e instance GenericVisibleEvent`, cause the intention of the program statistics is to give a rough understanding of all the visible events to gauge the size of the execution graphs.
Like this, you will miss visible non-fences (like the ones in Vulkan).

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorArm8.java
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorArm8.java:174]: I would rather use `Event` as variable type. The code doesn't really need to know about what type of event it is. I think we can do that for pretty much all cases where `GenericVisibleEvent` is used right now.

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusRISCV.java
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusRISCV.java:238]: you can use a switch expression:
```
Event fence = switch (...) {
  case "r.r" -> EventFactory.RISCV.newRRFence();
  case "r.w" -> ....
}
```

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java:110]: That would add `name` to the tag to every single event which is exactly what you initially wrote we should not do. I am having doubts about what would be the best approach. On the one side, I think it is better not to add the name by default and do it only if needed by creating a specific factory method for the special instruction and add the tag there (in like we are doing now in e.g., RISCV). However, except for PTX (and the soon coming vulkan), it seems pretty much all our memory models indeed require having the name as a tag.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/common/FenceBase.java
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/common/FenceBase.java:11]: No ... these are language level fences and I argue that for those `mo` is mandatory (which is aligned with the fact we have `Preconditions.checkArgument(!mo.isEmpty());` in the constructor)

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorArm8.java
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorArm8.java:174]: What about the factory? There we always return the most precise type so for consistency I would still use `GenericVisibleEvent`

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/common/FenceBase.java
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/common/FenceBase.java:11]: `Fence` does not exist anymore, hence it is outdated. Strictly speaking, this class is also not just for language-level fences, it is just a common(!) implementation that can be reused for whatever purposes.
Whether all fences have a `mo`or not is likely a matter of convention. Intuitively, I would have said a compiler barrier has no memory ordering, but in our `linux.h` we do model compiler barriers as a fence with `barrier` memory ordering. I'm fine with that, though I don't necessarily see why `FenceBase` should enforce that.
After all, this class is abstract and any subclass is free to decide if it needs a memory ordering or not.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java:110]: I said that it would not be a good idea to do it in the constructor of `GenericVisibleEvent` and that the `EventFactory` could do it, if necessary. Not doing it in the constructor allows you to easily support `EventFactory.newFence` which automatically adds name + F-tag, while also supporting something like `EventFactory.newVisibleEvent` which does not add anything at all. Honestly, we might even go a step further and just have `GenericEvent` that does not even need to be visible and let the `EventFactory` add the visible tag.
That being said, I understood that `PTX` and `Vulkan` just do not use the F-tag, so they have non-fence events.
So you wouldn't call `EventFactory.newFence` in the first place, whether it adds the name or not.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorArm8.java
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorArm8.java:174]: In the factory you can use `GenericVisibleEvent`. I would see it similar to how we use `Set<T>` as a type even if we instantiate it with `HashMap<T>` most of the time. If the concrete type is not so important at the call site, I would stick with the more generic type there.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java:110]: I am not referring to the new instructions from vulkan that gave raise to the change in this PR, but rather to PTX or vulkan **fences**. E.g. PTX has [this kind](https://github.com/hernanponcedeleon/Dat3M/blob/027b4024dd3b9894473233bc5f197912a7cf7421/litmus/PTX/Proxy-Alias-AliasFence.litmus#L10) of instructions which in the memory model are simply referred as `F` (different to e.g. `tso` which expliclty uses `mfence`)

Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java:110]: I see. Well, that is also the case for all C11 fences: the cat model always refers to `F & mo` never to the actual fence name which would be `atomic_thread_fence` I suppose. So there a few cases where the name of the fence is irrelevant.

Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java:110]: Exactly and this is why I don't want `newFence` to add it. Instead, we should call e.g., `newSYBarrier` and that one would add the needed tag.

Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java:110]: Ok, but you could still add a method like `newFenceWithoutNameTag` (you can even make it private if you want) to avoid the code repetition. I mean, we can easily provide different options with different factory methods.
Though I guess in most cases it does not really hurt to add the name as a tag as well, but I agree that we can avoid it entirely.

Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/EventFactory.java]: It seems pretty much all our MMs were relying on the name being add as a tag, either because the cat file refers to it (e.g., `mfence`) or because we about this and use it to pass the memory ordering (e.g., compilation of C11). Thus, I decided to keep the functionality as closed as possible to before.

File: dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java:642]: Can we use `List<Event>` and go over all(!) events and filter for those that have `Visible`, `Fence` and `PTX.SC` tag?

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusRISCV.java
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusRISCV.java:245]: Add the wrong fence mode as part of the error message, so it is easy to see what is wrong.
I would also say "Invalid fence mode" cause "no support/unsupported" sounds like we have an incomplete implementation (do we?).

File: dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java:980]: Here I would also go over all events and take those that have `Visible` and `Fence` tag.

File: dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java:1028]: Same as above. But you can additionally filter for `Tag.PTX.SC`, because the loop only ever cares about those events.

**Review by ThomasHaas — COMMENTED** [review]
Apart from the remaining few and minor issues, the PR looks good to me. I will approve once the changes are there

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusRISCV.java
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusRISCV.java:245]: We are complete. We even support the synchronization fence `.i` which is used in several litmus tests, but it is really not part of the cat file (I noticed this the other day). I double checked both in herd's repo and RISCV documentation, and indeed this fence is not part of the formal memory model.

**Review by ThomasHaas — APPROVED** [review]
LGTM (I pushed some minor changes so you might want to double check before merging).

File: dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java
Comment (hernanponcedeleon) [inline dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java:641]: why do we need the sort?

Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java:641]: The original code used a sort, so I kept it. I think it is not necessary for correctness, but it makes sense to go over the events in that order.
That being said, the call to `program.getThreadEventsWithAllTags` should already return a sorted list, so the sort will do nothing (IIRC, java's sort is optimized to run in linear time on already sorted lists).
You can remove the call, if you want.

Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java:641]: Yes, the whole things boils down to call `e.getSuccessors()` so the ordering should be guaranteed (and even if not, it should not affect correctness). I will remove

**Pull request merged**
